# Requirements

Most analyses of the tutorials of the Phylogenomics course will be conducted using the lynx server of the Matschiner research group at LMU Munich, where the required tools are installed. There are a few exceptions, however, where the use of Graphical User Interface (GUI) programs will be required or beneficial; these should therefore be installed locally.

## Graphical User Interface programs

* **A console program:** A console is essential for work on servers like lynx. On MacOSX, the standard console program is called [Terminal](https://en.wikipedia.org/wiki/Terminal_(macOS)). On Linux, several options are available, including [Konsole](https://en.wikipedia.org/wiki/Konsole). If you're a Linux user, you're likely familiar with console programs. On recent Windows systems, a console program is also installed and should be called [Windows Terminal](https://en.wikipedia.org/wiki/Windows_Terminal).

* **AliView:** To visualize sequence alignments, the software [AliView](http://www.ormbunkar.se/aliview/) ([Larsson 2014](https://academic.oup.com/bioinformatics/article/30/22/3276/2391211)) is recommended. The installation of AliView is described at [http://www.ormbunkar.se/aliview/](http://www.ormbunkar.se/aliview/) and should be possible on all operating systems.

* **BEAST2:** The BEAST2 package, including BEAUti, BEAST2 itself, TreeAnnotator, and other tools can be downloaded from the BEAST2 website [https://www.beast2.org](https://www.beast2.org). As all these programs are written in Java, compilation is not required, and all programs should work on Mac OS X, Linux, and Windows.

* **Tracer:** The program [Tracer](http://beast.community/tracer) greatly facilitates the inspection of output from Bayesian analyses such as those done with BEAST2. It is a GUI program that therefore can not be used on lynx, but it is easy to install on your local computer. Input files for Tracer will thus need to be downloaded from lynx after BEAST2 analyses. Executables of Tracer for MacOS, Linux, and Window can be found on [https://github.com/beast-dev/tracer/releases](https://github.com/beast-dev/tracer/releases).

* **FigTree:** The program [FigTree](http://tree.bio.ed.ac.uk/software/figtree/) is a very intuitive and useful tool for the visualization and (to a limited extent) manipulation of phylogenies encoded in [Newick](http://evolution.genetics.washington.edu/phylip/newicktree.html) format. Being a GUI program, FigTree can not be run on lynx, but needs to be installed and used on your local computer. Input files for FigTree will thus need to be downloaded from lynx. Executables of FigTree for Mac OS X, Linux, and Windows are provided on [https://github.com/rambaut/figtree/releases](https://github.com/rambaut/figtree/releases). Download the file [FigTree.v1.4.4.dmg](https://github.com/rambaut/figtree/releases/download/v1.4.4/FigTree.v1.4.4.dmg) if your local computer is running MacOS, [FigTree_v1.4.4.tgz](https://github.com/rambaut/figtree/releases/download/v1.4.4/FigTree_v1.4.4.tgz) if it is running Linux, and [FigTree.v1.4.4.zip](https://github.com/rambaut/figtree/releases/download/v1.4.4/FigTree.v1.4.4.zip) if it is running Windows.

* **RStudio:** [RStudio](https://www.rstudio.com/products/rstudio/) is an integrated development environment (IDE) for [R](https://www.r-project.org/). It includes a console, syntax-highlighting editor that supports direct code execution, as well as tools for plotting, history, debugging and workspace management. RStudio is available in open source and commercial editions and runs on the desktop (Windows, Mac, and Linux) or in a browser connected to RStudio Server or RStudio Server Pro (Debian/Ubuntu, Red Hat/CentOS, and SUSE Linux). Download the program [RStudio Desktop](https://www.rstudio.com/products/rstudio/download/). You have to choose the free version, which is not for the server. You will also have to install [R 3.6.0+](https://cran.rstudio.com/) on your computer before you install R Studio. Choose the OS appropriate for your computer.
